### PR TITLE
Getters

### PR DESCRIPTION
**Pull request by nisabek**
- changed X509_NAME_ENTRY fields to support list of strings
- moved some code to separate functions
- added make commands to run perf (sofar only for crypto folder) and view it 
- added conversion from PEM to cCert - WIP 

@spirius  - can you please review the TODO-s for you? they are marked with @spir :)

Thanks!

### CODE REVIEW

File: crypto/cert.go [hunk 1]
```diff
@@ -161,11 +140,87 @@ func CreateCertificate(certRequest *CertificateRequest) (*Certificate, error) {
...
+        return nil, GoifyError("could not sign %s")
     }
     return &cert, nil
 }
 
+/**
+    To be used later for internal calls (no need to create certificate request then)
+ */
+func createSubjectNameFromMap(nameEntryMap map[NID][]string) (*Name, error) {
+    name := &Name{}
```
Comment (spirius) [inline crypto/cert.go]: Please make a separate function to create Name type struct with finalizer to be used in function createSubjectNameFromMap and createSubjectName.

File: crypto/cert.go [hunk 2]
```diff
@@ -198,18 +253,54 @@ func setSerial(cCert *C.X509, serial uint64) error {
...
-        bytes := []byte(value)
-        ret := C.X509_NAME_add_entry_by_NID(cName, C.int(nid), C.MBSTRING_UTF8, (*C.uchar)(unsafe.Pointer(&bytes[0])), C.int(len(value)), -1, 0)
+        for _, value := range values {
+            bytes := []byte(value)
+            ret := C.X509_NAME_add_entry_by_NID(cName, C.int(nid), C.MBSTRING_UTF8, (*C.uchar)(unsafe.Pointer(&bytes[0])), C.int(len(value)), -1, 0)
 
-        if ret != 1 {
-            err = fmt.Errorf("could not add name entry NID: %d, result: %d, %s", nid, ret, C.GoString(C.ERR_error_string(C.ERR_get_error(), nil)))
+            if ret != 1 {
+                err = fmt.Errorf("could not add name entry NID: %d, result: %d, %s", nid, ret, C.GoString(C.ERR_error_string(C.ERR_get_error(), nil)))
```
Comment (spirius) [inline crypto/cert.go]: break in case of error?

File: crypto/cert.go [hunk 3]
```diff
@@ -257,3 +348,124 @@ func AddExtension(cert *Certificate, nid NID, value string, err error) (error) {
...
+func ReadCertificate(cert *Certificate, pem string) (error) {
+    bio := C.BIO_new(C.BIO_s_mem())
+    if bio == nil {
+        return GoifyError("Could not initialize BIO memory: %s")
+    }
+
+    defer C.BIO_free(bio)
+
+    rc := C.BIO_puts(bio, C.CString(pem))
+    if rc <= 0 { //TODO @spir... A 0 or -1 return is not necessarily an indication of an error. In particular when the source/sink is non-blocking or of a certain type it may merely be an indication that no data is currently available and that the application should retry the operation later.
```
Comment (spirius) [inline crypto/cert.go]: we are not using non-blocking IO, so data should be readable/available immediately

File: crypto/cert.go [hunk 4]
```diff
@@ -257,3 +348,124 @@ func AddExtension(cert *Certificate, nid NID, value string, err error) (error) {
...
+        return GoifyError("could not write string into BIO: %s")
+    }
+
+    cert.cCert = C.PEM_read_bio_X509(bio, nil, nil, nil)
```
Comment (spirius) [inline crypto/cert.go]: error checking?

File: crypto/cert.go [hunk 5]
```diff
@@ -257,3 +348,124 @@ func AddExtension(cert *Certificate, nid NID, value string, err error) (error) {
...
+
+    return nil;
+}
+
+/**
+    Gets the commonName (all available) from the subjectName X509_NAME object
+
+    TODO @spir - make setters also implement Certificate?..
```
Comment (spirius) [inline crypto/cert.go]: yes, would be more consistent

File: crypto/cert.go [hunk 6]
```diff
@@ -257,3 +348,124 @@ func AddExtension(cert *Certificate, nid NID, value string, err error) (error) {
...
+        lastpos = C.X509_NAME_get_index_by_NID(cSubjectName, C.int(NID_commonName), lastpos);
+        if lastpos == -1 {
+            break;
+        }
+        cEntry := C.X509_NAME_get_entry(cSubjectName, lastpos);
+
+        name, err := goStringFromNameEntry(cEntry)
+
+        if err != nil {
+            return commonName, err //todo @spir - return as much as I could read?
```
Comment (spirius) [inline crypto/cert.go]: yep, looks good

**Comment by spirius** [comment]
the rest looks good

File: crypto/cert.go [hunk 7]
```diff
@@ -257,3 +348,124 @@ func AddExtension(cert *Certificate, nid NID, value string, err error) (error) {
...
+
+    defer C.free(strMem)
+
+    rc := C.ASN1_STRING_to_UTF8((**C.uchar)(unsafe.Pointer(&strMem)), entryStr);
+
+    if rc <= 0 {
+        return "", GoifyError("could not convert ASN1_string")
+    }
+
+    bufPointer := unsafe.Pointer(strMem) // todo @spir - do I need to do this?..
```
Comment (nisabek) [inline crypto/cert.go]: @spirius  - what about this one?

**Pull request merged**